### ISSUE TITLE
Collect stats about message sizes in JITServer

### ISSUE THREAD

**Issue by mpirvu** [eclipse/openj9#9708]
The goal is to determine which messages are larger and whether such messages can be reduced in size. For instance, the message that sends the compilation request is probably on the large side because it sends the entire ROMClass to the JITServer. If we find it worthwhile, we may be able to keep some cache at the client that memorizes which ROMClasses have already been sent to the server so that we don't have to send them again.

**Comment by mpirvu** [comment]
I am thinking that the best place for collecting message sizes is:
```
void
CommunicationStream::readMessage(Message &msg)
   {
   msg.clearForRead();

   // read message size
   uint32_t serializedSize;
   readBlocking(serializedSize);
   msg.setSerializedSize(serializedSize);

   // read the rest of the message
   uint32_t messageSize = serializedSize - sizeof(uint32_t);
   readBlocking(msg.getBufferStartForRead() + sizeof(uint32_t), messageSize);

   // rebuild the message
   msg.deserialize();
   }
```
Here we have access to the total message size and the type of the message can be obtained with `msg.type()`

**Comment by mpirvu** [comment]
Attn: @EmanElsaban

**Comment by mpirvu** [comment]
Stats could be collected with utilities from `omr/compiler/infra/Statistics.hpp`
`class TR_Stats`
or `class TR_StatsHisto` when the spread is large

**Comment by EmanElsaban** [comment]
`JITServer Message Type Statistics:
Type# #called           Max             Min             Mean            StdDev          Sum             TypeName
#0000    4713   65204.000000    684.000000      3948.217271     4515.226255     18607948.000000         compilationCode
#0001      20   24.000000       24.000000       24.000000       0.000000        480.000000              compilationFailure
#0002    5914   68.000000       68.000000       67.988510       1.250129        402356.000000           mirrorResolvedJ9Method
#0003    3380   44.000000       44.000000       44.000000       0.000000        148720.000000           get_params_to_construct_TR_j9method
#0004       1   24.000000       24.000000       24.000000       0.000000        24.000000               getUnloadedClassRangesAndCHTable
#0014     236   64.000000       64.000000       64.000000       0.000000        15104.000000            ResolvedMethod_staticAttributes
#0015    4190   52.000000       52.000000       52.000000       0.000000        217932.000000           ResolvedMethod_getClassFromConstantPool
#0016   24380   40.000000       40.000000       40.000000       0.256126        975640.000000           ResolvedMethod_getDeclaringClassFromFieldOrStatic
#0017    2689   52.000000       52.000000       52.000000       0.000000        139932.000000           ResolvedMethod_classOfStatic
#0018     219   28.000000       28.000000       28.000000       0.000000        6132.000000             ResolvedMethod_startAddressForJNIMethod
#0019     764   64.000000       64.000000       64.000000       0.000000        48896.000000            ResolvedMethod_fieldAttributes
#0020    6377   40.000000       40.000000       40.000000       0.000000        255080.000000           ResolvedMethod_getResolvedStaticMethodAndMirror
#0021   20142   40.000000       40.000000       39.998014       0.281816        805800.000000           ResolvedMethod_getResolvedSpecialMethodAndMirror
#0022     130   40.000000       40.000000       40.000000       0.000000        5200.000000             ResolvedMethod_classCPIndexOfMethod
#0025   11916   56.000000       56.000000       56.000000       0.000000        667464.000000           ResolvedMethod_getResolvedPossiblyPrivateVirtualMe
#0027     350   40.000000       40.000000       40.000000       0.000000        14000.000000            ResolvedMethod_getResolvedInterfaceMethod_2
#0028   23903   76.000000       76.000000       75.996821       0.491562        1816628.000000          ResolvedMethod_getResolvedInterfaceMethodAndMirror
#0029     292   56.000000       56.000000       56.000000       0.000000        16352.000000            ResolvedMethod_getResolvedInterfaceMethodOffset
#0030    1493   40.000000       40.000000       40.000000       0.000000        59720.000000            ResolvedMethod_getUnresolvedStaticMethodInCP
#0031     209   28.000000       28.000000       28.000000       0.000000        5852.000000             ResolvedMethod_isSubjectToPhaseChange
#0032   11657   40.000000       40.000000       40.000000       0.000000        466360.000000           ResolvedMethod_getUnresolvedSpecialMethodInCP
#0033    2722   40.000000       40.000000       40.000000       0.000000        108880.000000           ResolvedMethod_getUnresolvedFieldInCP
#0034    9527   68.000000       60.000000       65.044500       3.861423        619744.000000           ResolvedMethod_getRemoteROMString
#0035     994   40.000000       40.000000       40.000000       0.000000        39760.000000            ResolvedMethod_fieldOrStaticName
#0036    1863   28.000000       28.000000       28.000000       0.000000        52164.000000            ResolvedMethod_getRemoteROMClassAndMethods
#0040      30   40.000000       40.000000       40.000000       0.000000        1200.000000             ResolvedMethod_isUnresolvedCallSiteTableEntry
#0041      29   40.000000       40.000000       40.000000       0.000000        1160.000000             ResolvedMethod_callSiteTableEntryAddress
#0042      10   44.000000       44.000000       44.000000       0.000000        440.000000              ResolvedMethod_getResolvedDynamicMethod
#0044       6   44.000000       44.000000       44.000000       0.000000        264.000000              ResolvedMethod_isSameMethod
#0054     164   40.000000       40.000000       40.000000       0.000000        6560.000000             ResolvedMethod_getResolvedImproperInterfaceMethodA
#0056    3879   40.000000       40.000000       40.000000       0.000000        155160.000000           ResolvedMethod_stringConstant
#0057    2207   68.000000       68.000000       68.000000       0.000000        150076.000000           ResolvedMethod_getResolvedVirtualMethod
#0058   11407   924.000000      60.000000       79.453717       37.931013       906408.000000           ResolvedMethod_getMultipleResolvedMethods
#0059      30   40.000000       40.000000       40.000000       0.000000        1200.000000             ResolvedMethod_varHandleMethodTypeTableEntryAddres
#0060      26   40.000000       40.000000       40.000000       0.000000        1040.000000             ResolvedMethod_isUnresolvedVarHandleMethodTypeTabl
#0064    5401   52.000000       52.000000       52.000000       0.000000        280852.000000           ResolvedMethod_definingClassFromCPFieldRef
#0066    4689   64.000000       64.000000       64.000000       0.000000        300160.000000           ResolvedRelocatableMethod_fieldAttributes
#0067    3106   64.000000       64.000000       64.000000       0.000000        198784.000000           ResolvedRelocatableMethod_staticAttributes
#0076     107   56.000000       48.000000       54.205607       2.760019        5800.000000             VM_getSystemClassFromClassName
#0078       1   28.000000       28.000000       28.000000       0.000000        28.000000               VM_getClassClassPointer
#0079     998   332.000000      60.000000       120.577154      39.959703       120336.000000           VM_setJ2IThunk
#0084    7150   152.000000      56.000000       93.016361       14.623389       665160.000000           VM_getClassFromSignature
#0085   16135   84.000000       84.000000       83.989588       1.322589        1355172.000000          VM_jitFieldsAreSame
#0086   14223   72.000000       72.000000       72.000000       0.000000        1024056.000000          VM_jitStaticsAreSame
#0094       5   28.000000       28.000000       28.000000       0.000000        140.000000              VM_isClassInitialized
#0097      27   28.000000       28.000000       28.000000       0.000000        756.000000              VM_isString1
#0102       5   24.000000       24.000000       24.000000       0.000000        120.000000              VM_stackWalkerMaySkipFrames
#0108      19   24.000000       24.000000       24.000000       0.000000        456.000000              VM_getClassFromNewArrayType
#0111       8   28.000000       28.000000       28.000000       0.000000        224.000000              VM_getArrayClassFromComponentClass
#0121     147   24.000000       24.000000       24.000000       0.000000        3528.000000             VM_getOffsetOfClassFromJavaLangClassField
#0127       1   52.000000       52.000000       52.000000       0.000000        52.000000               VM_getResolvedVirtualMethod
#0130     244   100.000000      84.000000       92.000000       7.983522        22448.000000            VM_getInstanceFieldOffset
#0133     165   28.000000       28.000000       28.000000       0.000000        4620.000000             VM_getClassDepthAndFlagsValue
#0134     241   172.000000      72.000000       104.763485      26.162594       25248.000000            VM_getMethodFromName
#0142      24   44.000000       44.000000       44.000000       0.000000        1056.000000             VM_getInterpreterVTableSlot
#0154      42   64.000000       44.000000       60.476190       4.435104        2540.000000             VM_setInvokeExactJ2IThunk
#0155     109   28.000000       28.000000       28.000000       0.000000        3052.000000             VM_createMethodHandleArchetypeSpecimen
#0156       9   44.000000       44.000000       44.000000       0.000000        396.000000              VM_instanceOfOrCheckCast
#0157       4   28.000000       28.000000       28.000000       0.000000        112.000000              VM_getResolvedMethodsAndMirror
#0158       1   24.000000       24.000000       24.000000       0.000000        24.000000               VM_getVMInfo
#0160      15   40.000000       40.000000       40.000000       0.000000        600.000000              VM_dereferenceStaticAddress
#0164    1080   284.000000      24.000000       77.768733       37.469323       84068.000000            VM_getJ2IThunk
#0165      42   216.000000      28.000000       81.333333       38.838419       3416.000000             VM_needsInvokeExactJ2IThunk
#0193    1902   40.000000       40.000000       40.000000       0.000000        76080.000000            SharedCache_rememberClass
#0194       3   40.000000       40.000000       40.000000       0.000000        120.000000              SharedCache_addHint
#0196       4   40.000000       40.000000       40.000000       0.000000        160.000000              runFEMacro_invokeCollectHandleNumArgsToCollect
#0197      17   28.000000       28.000000       28.000000       0.000000        476.000000              runFEMacro_invokeExplicitCastHandleConvertArgs
#0198      49   40.000000       40.000000       40.000000       0.000000        1960.000000             runFEMacro_targetTypeL
#0199     105   44.000000       36.000000       38.514286       3.731658        4044.000000             runFEMacro_invokeILGenMacrosInvokeExactAndFixup
#0200      18   28.000000       28.000000       28.000000       0.000000        504.000000              runFEMacro_invokeArgumentMoverHandlePermuteArgs
#0201      18   28.000000       28.000000       28.000000       0.000000        504.000000              runFEMacro_invokePermuteHandlePermuteArgs
#0204       3   52.000000       52.000000       52.000000       0.000000        156.000000              runFEMacro_invokeDirectHandleDirectCall
#0207      33   28.000000       28.000000       28.000000       0.000000        924.000000              runFEMacro_invokeFoldHandle
#0208      66   28.000000       28.000000       28.000000       0.000000        1848.000000             runFEMacro_invokeFoldHandle2
#0210       3   40.000000       40.000000       40.000000       0.000000        120.000000              runFEMacro_invokeFilterArgumentsHandle
#0211      12   28.000000       28.000000       28.000000       0.000000        336.000000              runFEMacro_invokeFilterArgumentsHandle2
#0224   13993   60.000000       60.000000       60.000000       0.000000        839640.000000           IProfiler_profilingSample
#0225    4590   28.000000       28.000000       28.000000       0.000000        128548.000000           IProfiler_searchForMethodSample
#0227    4764   52.000000       52.000000       52.000000       0.000000        247728.000000           IProfiler_setCallCount
#0232      77   28.000000       28.000000       28.000000       0.000000        2156.000000             KnownObjectTable_getOrCreateIndexAt
#0234     113   28.000000       28.000000       28.000000       0.000000        3164.000000             KnownObjectTable_getExistingIndexAt
#0235     125   56.000000       56.000000       56.000000       0.000000        7000.000000             KnownObjectTable_symbolReferenceTableCreateKnownOb
#0237       3   68.000000       68.000000       68.000000       0.000000        204.000000              KnownObjectTable_dereferenceKnownObjectField
#0241       7   40.000000       40.000000       40.000000       0.000000        280.000000              KnownObjectTable_invokeDirectHandleDirectCall
Total number of messages: 235772. Average number of messages per compilation:50.025887
`

**Comment by EmanElsaban** [comment]
[AcmeAir Message Statistics.xlsx](https://github.com/eclipse/openj9/files/4747286/AcmeAir.Message.Statistics.xlsx)

**Comment by mpirvu** [comment]
Thanks. 
These are the messages sent by the server and received by the client. It appears that the message with the compiled code can be pretty big, 1K-65K, but the rest are typically less than 100 bytes.

@EmanElsaban could you please generate the same stats from the server point of view (messages received by the server)? In this case you may need to add the print routine to another place.

**Comment by EmanElsaban** [comment]
I included both on the client and server side in this excel sheet, Thank you.
[AcmeAir MsgStats on client and server side.xlsx](https://github.com/eclipse/openj9/files/4754314/AcmeAir.MsgStats.on.client.and.server.side.xlsx)

**Comment by EmanElsaban** [comment]
I included the client and server side in this excel sheet, Thank you.
[AcmeAir MsgStats on client and server side.xlsx](https://github.com/eclipse/openj9/files/4754407/AcmeAir.MsgStats.on.client.and.server.side.xlsx)

**Comment by EmanElsaban** [comment]
I ran it a second trial on Acmeair to print out the total number of messages on server side:
[AcmeAir MsgStats on client and server side.xlsx](https://github.com/eclipse/openj9/files/4754494/AcmeAir.MsgStats.on.client.and.server.side.xlsx)

**Comment by mpirvu** [comment]
The compilation request is the largest contributor: it has the largest messages (115KB), the latest average message size (17.5 KB) and the largest number of bytes being received (95 MB). This represents 55% of the total bytes transferred from client to server.
Given this, it may be worthwhile to look into reducing the size of this message as mentioned in the first comment of this issue

**Comment by mpirvu** [comment]
Resolved by #9870
